### PR TITLE
Add a hermetic PipeListener for testing net code

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
I added a `PipeListener` utility that will make it easier to run tests involving net.Dialer (which is a lot of our tests). I plan to use this in tests for the `membership` package.

**Why?**
To make writing net code tests easier, faster and safer.

**How did you test it?**
This package has 100% test coverage including goleak testing. In addition, net.Pipe is already commonly used for this purpose. I just couldn't find a good library that already does this, and does it properly.

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: internal/nettest/fx.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+var Module = fx.Options(
+	fx.Provide(func(c clock.Clock) []Option {
+		return []Option{func(params *PipeListenerParams) {
+			params.Clock = c
+		}}
+	}),
+	fx.Provide(func(options []Option) *PipeListener {
+		return NewPipeListener(options...)
+	}),
+	fx.Invoke(func(lifecycle fx.Lifecycle, l *PipeListener) {
```
Comment (dnr) [inline internal/nettest/fx.go:45]: Only one listener? What happens if a test needs more?

**Review by dnr — COMMENTED** [review]
We've been doing okay with localhost tcp connections.. is this really going to make things significantly faster/more reliable? Specifically I'm a little concerned about replicating various edge-case tcp behavior (which usually doesn't matter, except when it does)

Do you have an example PR that rewrites a test using this? Or a new test?

File: internal/nettest/fx.go (hunk 1 above)
Comment (MichaelSnowden) [reply internal/nettest/fx.go]: Good point. I made this provide `PipeListenerFactory` instead.

File: internal/nettest/fx.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// Module provides a PipeListenerFactory with a clock.Clock as the sole dependency.
+var Module = fx.Options(
+	fx.Provide(func(c clock.Clock) []Option {
+		return []Option{func(params *PipeListenerParams) {
+			params.Clock = c
+		}}
+	}),
+	fx.Provide(func(options []Option) PipeListenerFactory {
```
Comment (dnr) [inline internal/nettest/fx.go:50]: Do we really need a factory and to involve fx here? We already have `NewPipeListener` exposed at the package level

**Comment by dnr** [comment]
> > We've been doing okay with localhost tcp connections.. is this really going to make things significantly faster
> 
> Good question. I just added a benchmark test which compares this to TCP and Unix sockets. In both instances, this listener is roughly ~10x faster.

That's nice but it doesn't answer the relevant question, which is how much faster does this make a suite of our unit tests that use this compared to the same suite using tcp sockets. My guess is it's not noticeable, y'know, Amdahl's law and all. It still might be nice for the reliability factor though

**Review by dnr — APPROVED** [review]

File: internal/nettest/fx.go (hunk 2 above)
Comment (MichaelSnowden) [reply internal/nettest/fx.go]: No, probably not. I just removed it.

**Comment by MichaelSnowden** [comment]
I simplified this a lot in the latest update. It's a lot smaller now.
